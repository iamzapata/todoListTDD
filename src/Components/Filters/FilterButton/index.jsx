import React, { PureComponent } from "react"
import { func, string, node, bool } from "prop-types"

class FilterButton extends PureComponent {
  render() {
    const {
      filter,
      setFilter,
      isListEmpty,
      children,
      currentFilter
    } = this.props
    return (
      <button
        onClick={() => setFilter(filter)}
        disabled={isListEmpty || filter === currentFilter}
      >
        {children}
      </button>
    )
  }
}

FilterButton.propTypes = {
  filter: string.isRequired,
  setFilter: func.isRequired,
  children: node.isRequired,
  isListEmpty: bool.isRequired,
  currentFilter: string.isRequired
}

export default FilterButton
