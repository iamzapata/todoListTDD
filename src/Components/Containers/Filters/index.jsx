import React, { PureComponent } from "react"
import { func, bool } from "prop-types"
import { connect } from "react-redux"
import splitToLowercase from "../../../utils/splitToLowerCase/index"
import FilterButton from "../../Filters/FilterButton/index"
import { isListEmpty } from "../../../reducers/TodosStore"
import { setFilter } from "../../../actions/filters"

export class Filters extends PureComponent {
  render() {
    const { setFilter, isListEmpty, currentFilter } = this.props
    return (
      <div className="Filters">
        <label>Show:</label>
        {["SHOW_ALL", "SHOW_ACTIVE", "SHOW_COMPLETED"].map(filter => (
          <FilterButton
            key={filter}
            filter={filter}
            setFilter={setFilter}
            isListEmpty={isListEmpty}
            currentFilter={currentFilter}
          >
            {splitToLowercase(filter)("_")}
          </FilterButton>
        ))}
      </div>
    )
  }
}

Filters.propTypes = {
  setFilter: func,
  isListEmpty: bool
}

const mapStateToProps = state => ({
  isListEmpty: isListEmpty(state),
  currentFilter: state.FiltersStore.filter
})

const mapDispatchToProps = {
  setFilter
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filters)
