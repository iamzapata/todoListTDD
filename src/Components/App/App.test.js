import React from "react"
import { render } from "react-testing-library"
import { App } from "./index.jsx"
import { Filters } from "../Containers/Filters"
import { Provider } from "react-redux"

describe("<App />", () => {
  const store = {
    getState: () => ({
      TodosStore: {
        todos: []
      },
      FiltersStore: {
        filter: ""
      }
    }),
    subscribe: jest.fn(),
    dispatch: jest.fn()
  }
  function renderWithRedux(ui) {
    return {
      ...render(<Provider store={store}>{ui}</Provider>)
    }
  }

  const createProps = () => ({
    title: "Hello World !",
    todos: [],
    addTodo: jest.fn(),
    toggleTodo: jest.fn()
  })

  it("Should render self", () => {
    const props = createProps()
    const { container, queryByText } = renderWithRedux(<App {...props} />)
    const header = queryByText(props.title)
    expect(container.firstChild).toMatchSnapshot()
    expect(header.innerHTML).toBe(props.title)
  })

  it("Should contain a <TodosList /> component", () => {
    const { container } = renderWithRedux(<App {...createProps()} />)
    const TodosList = container.querySelectorAll(".TodosList")

    expect(TodosList.length).toEqual(1)
  })

  it("Should contain a <TodosInput /> component", () => {
    const { container } = renderWithRedux(<App {...createProps()} />)
    const TodosInput = container.querySelectorAll(".TodoInput")

    expect(TodosInput.length).toEqual(1)
  })

  it("Should have <TodosList /> bellow <TodosInput />>", () => {
    const { container } = renderWithRedux(<App {...createProps()} />)
    const elementPositioning = container.querySelector(
      ".TodoInput + .TodosList"
    )

    expect(elementPositioning).toBeTruthy()
  })

  it("Should contain a <Filters /> component", () => {
    const { container } = renderWithRedux(<App {...createProps()} />)
    const Filters = container.querySelectorAll(".Filters")

    expect(Filters.length).toBeTruthy()
  })
})
