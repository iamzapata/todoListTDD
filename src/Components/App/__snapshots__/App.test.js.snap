// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should render self 1`] = `
<h1>
  Hello World !
</h1>
`;
